BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+BuildXL as a library - aka. BXLLib.
+
+Problem statement
+-----------------
+- Performance
+  - Graph construction takes a long time due to DScript parsing, type checking, and evaluation.
+  - DScript frontend is not scalable to handle Office multi-architecture (mularchy) builds.
+
+- Frontend usability
+  - CMake, MsBuild frontends create their own adhoc JSON formats of build graphs.
+
+- Binary graph fragments are proposed as a way to construct build graphs, but need a way (API)
+  to represent and construct those fragments.
+
+Requirements
+------------
+- API must preserve compatibility, i.e., maintain backward compatibility.
+  - API must be binary compatible.
+  - Graph fragment must be versioned.
+
+- Graph fragment construction must be high performant.
+  - Creating, serializing, and deserializing pips must be performant.
+  - Stitching multiple fragments must be performant.
+
+- Graph fragment interactions:
+  - Fragment can consume artifacts produced by other frontends.
+  - Fragment can produce artifacts that can be consumed by other frontends.
+
+- API for constructing graph fragments that makes sense for customer
+  - Adding basic pips, e.g., process, copy, write, seal directory.
+  - Extensible for adding BuildXL's friends:
+    - Adding QTest pips.
+    - Adding drop and symbol pips.    
+  - Desirable:
+    - Higher-level API to ease migration from existing DScript SDK.
+      - Creating a unique output directory.
+      - Specifying inputs/outputs of pips like in DScript.
+    - API becomes yet another frontend.
+
+- Reuse existing DScript (or other frontends) SDK: Drop, Symbol, QTest, etc.
+
+- Control over BuildXL's friends behavior, e.g., adding a new default to QTest pips.
+
+- Track what files and environment variables used for graph fragment constructions.
+  - This is needed in case the construction is not done in a sandboxed environment.
+
+- BXLLib is applicable to many customers, not only to Office, e.g., WDG can benefit from BXLLib.
+
+Spike
+-----
+- Use BuildXL DScript frontend to evaluate specs and generate specific fragments.
+  - Description:
+    - This approach targets only performance.
+    - In metabuild, besides generating DScript spec, the pip also evaluates the spec (plus the SDK) and generates a fragment.
+    - Instead of using a full-blown BuildXL, we refactor BuildXL frontend to create so-called BXLScriptAnalyzer
+      that is used for evaluation and for generating fragment.

```

Comment:
Do the BXLScriptAnalyzer pips need to run in topological order with respect to the dependencies between graph fragments?